### ISSUE TITLE
RefCount variable right after IUnknown interface.

### ISSUE THREAD

**Issue by Kr00l** [WaynePhillipsEA/twinbasic#433]
**Describe the bug**
In VBx there is a hidden RefCount variable in every COM object stored right after the IUnknown interface.
tB seems not to store that RefCount variable in that location. It may be helpful to make the same:
1. to be 100% compatible
2. to allow a (hidden) way to retrieve the value of how many references a COM object currently has.

**To Reproduce**
```
Class Class1
    'Void
End Class

Module MainModule
    
    Option Explicit
    Private Declare Sub CopyMemory Lib "kernel32" Alias "RtlMoveMemory" (ByRef Dest As Any, ByRef Source As Any, ByVal Bytes As Long)

    Public Sub Main()
        Dim This As Class1
        Set This = New Class1
        MsgBox GetRefCount(This)
    End Sub
    
    Public Function GetRefCount(ByRef Object As IUnknown) As Long
        Dim RefCount As Long
        If Object Is Nothing Then
            Exit Function
        Else
            CopyMemory RefCount, ByVal (ObjPtr(Object) + 4&), 4
            GetRefCount = RefCount - 2 ' minus 2 to consider reference increase for function argument and ObjPtr on API call.
        End If
    End Function
	
End Module
```

**Expected behavior**
VBx returns 1 in above example where tB returns -2.

**Screenshots**
n/a

**Desktop (please complete the following information):**
 - OS: [e.g. Windows 10]
 - twinBASIC compiler version [e.g. 0.10.5788]

**Additional context**
n/a

**Comment by Greedquest** [comment]
That's an implementation detail. It's debatable whether tB should stick to these, I imagine some implementation details will be in direct conflict with new features e.g. inheritance and function overloads. Maybe not this one but I'm sure cases will arise.

I think what's more important is to discuss the features that such low level hacks provide workarounds for, and instead find a way to include those in tB properly. Why do you need the refcount? A debugger? Weak references? These are things that may be possible to request native support for instead.

**Comment by Kr00l** [comment]
Yes it's for debugging. Sometimes a object don't terminate in a complex scenario. This is a way to monitor it.
I mean tB needs the RefCount as it's a COM Law. It's a detail where it is stored. And to store it right next to IUnknown should be very easy and no harm for any future features.

**Comment by bclothier** [comment]
I'm wondering if really "every COM object" implements this way. It might be an artifact of how VBx arranges its COM interfaces, maybe MFC/ATL, maybe all Microsoft's provided objects, but we have no guarantees that *all* COM objects will have that variable in the same location. 

That would then yield garbage output.

So it's already fragile. And if I understand it, this isn't used in production; only as a debugging aid. I'm then more inclined to modify this request into a request to make it possible to track the ref count natively with the VSCode debugger, which can be reliably & safely obtained by doing a `IUnknown::AddRef`, capturing the return value, then subtracting 1, then calling `IUnknown::Release` on the COM object. No hacks required.

**Comment by Kr00l** [comment]
Okay, thanks bclothier.
I replaced now the "hacky" (or stupid) approach with below one and it works of course in VBx and in tB.
```
Public Function GetRefCount(ByRef Object As OLEGuids.IUnknownUnrestricted) As Long
    GetRefCount = Object.AddRef() - 2
    Object.Release
End Function
```

I leave it up to Wayne to choose what happens with this request.

**Comment by WaynePhillipsEA** [comment]
Well I think we should certainly offer a built-in GetObjRefCount function off of the VBA.[_HiddenModule] module (next to ObjPtr, VarPtr etc).   I already have this function in my Debug builds of tB and I use it in my unit tests a lot.

In addition, perhaps we should also report the RefCount in the debugger variables list (as the last field member of each object).  We might want a project level switch for that though(?), and perhaps default it to off.

**Comment by WaynePhillipsEA** [comment]
As for the actual placement of the RefCount field, it would currently be at -12 offset to the IUnknown.  Though in tB the root IUnknown vtable is appended to the end of the object, so its placement is not actually at a fixed position.   

A lot of effort went into ensuring the vtable layout and field layout matches VBx (with all the quirks), but not the actual class object layout.  We can certainly look at matching the VBx class object layout in due course, as that will offer the best possible compatibility for when these types of hacks have been used in real-world code.

**Comment by Kr00l** [comment]
I feel already comfortable with the replacement. :D
However, I think that "hack" is maybe more often used as I got it originally from aboutvb (german site).. and that was a tip from year 2000.

Concerning naming suggestion ObjRefCount() sounds IMO better than GetObjRefCount() as ObjPtr() is not named GetObjPtr() either. (To group Obj* together)

**Comment by WaynePhillipsEA** [comment]
Yup, I'd agree with that :)
